🐞 bug: WDA signing is ambiguous when multiple identities share a display name

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Appium Version

3.6.0

### Appium Host Type

Other

### Are you using Appium components not maintained by the Appium team?

No

### Is this issue reproducible using the latest components?

- [ ] I confirm the issue is still reproducible with the latest component versions

### Current Behavior

When starting an XCUITest session on an iOS real device, WebDriverAgent fails during the build or installation step.

The captured log shows the following sequence:

1. The WDA project runs the `Embed app icon into Runner.app` post-action.
2. Code signing reports that an Apple Development identity is ambiguous because multiple matching identities are present in the login keychain.
3. The WDA runner fails signature verification or installation.
4. Appium reports that `xcodebuild` failed with exit code 65.

The affected component versions are:

- Appium 3.6.0
- appium-xcuitest-driver 11.17.7
- appium-webdriveragent 15.1.6
- Xcode SDK: iphoneos26.5
- Device OS: iOS 17.5.1
- Real device

### Expected Behavior

WebDriverAgent should build and install successfully when multiple code-signing identities have the same display name.

The post-action should use the exact signing identity or certificate hash selected for the current build. If the identity cannot be determined safely, the build should fail with a clear diagnostic instead of producing an invalidly signed runner.

### Appium Log

See the attached sanitized log in Further Information.

### Environment Details

- Debug details (`appium --show-debug-info`): Not available
- Last component version(s) which did not exhibit the problem:
  - Appium 3.4.2
  - appium-xcuitest-driver 11.4.2
  - appium-webdriveragent 12.2.2
- Platform and version of device under test: iOS 17.5.1
- Real device or emulator/simulator: Real device
- Xcode: 26.5 (17F42)
- SDK: iphoneos26.5
- macOS: 26.5.1 (25G76)
- Command Line Tools: 26.6.0.0.1781586589

### Minimal Reproducible Example

The failure depends on the host keychain containing multiple valid Apple Development identities with the same display name.

Reproduction steps:

1. On macOS with Xcode 26.5 (17F42), install Appium 3.6.0.
2. Install `appium-xcuitest-driver` 11.17.7. This driver resolves WebDriverAgent 15.1.6.
3. Connect a real iOS device and use the same signing configuration as the affected environment.
4. Keep multiple matching Apple Development identities with the same display name in the login keychain.
5. Start a new XCUITest session with capabilities similar to:

```json
{
  "platformName": "iOS",
  "appium:automationName": "XCUITest",
  "appium:udid": "<device-udid>",
  "appium:deviceName": "<device-name>",
  "appium:platformVersion": "17.5.1",
  "appium:bundleId": "<installed-app-bundle-id>",
  "appium:showXcodeLog": true,
  "appium:xcodeOrgId": "<team-id>",
  "appium:xcodeSigningId": "iPhone Developer",
  "appium:updatedWDABundleId": "<wda-bundle-id>"
}
```

The WDA build fails while running the `Embed app icon into Runner.app` post-action. Code signing reports that the signing identity is ambiguous, followed by WDA signature verification or installation failure.

### Further Information

The affected WDA script is:

https://github.com/appium/WebDriverAgent/blob/master/Scripts/embed-runner-icon.sh

The app-icon post-action was introduced by:

https://github.com/appium/WebDriverAgent/pull/1138

Apple documents that a certificate SHA-1 hash should be used when multiple identities have the same name:

https://developer.apple.com/documentation/technotes/tn3161-inside-code-signing-certificates

Sanitized Appium log:

```text
# Focused sanitized excerpt from the Appium console log
# The same WDA startup failure was repeated during the session retries.

[Environment]
Appium: 3.6.0
XCUITest driver: 11.17.7
WebDriverAgent: 15.1.6
Xcode: 26.5 (17F42)
SDK: iphoneos26.5
Device OS: iOS 17.5.1

[Session]
2026-08-13 13:28:42:844 i [AppiumDriver] Appium v3.6.0 creating new XCUITestDriver (v11.17.7) session
2026-08-13 13:28:42:876 i [XCUITestDriver] Normalized platformVersion capability value '17.5.1' to '17.5'
2026-08-13 13:28:42:878 i [XCUITestDriver] Using WDA agent: '<PATH>'

[WDA build]
2026-08-13 13:28:46:324 d [XCUITestDriver] Beginning test with command 'xcodebuild build-for-testing test-without-building -project <PATH> -scheme WebDriverAgentRunner -derivedDataPath <PATH> -destination id=<DEVICE_UDID> IPHONEOS_DEPLOYMENT_TARGET=17.5 DEVELOPMENT_TEAM=<TEAM_ID> CODE_SIGN_IDENTITY=iPhone Developer PRODUCT_BUNDLE_IDENTIFIER=<BUNDLE_ID> GCC_TREAT_WARNINGS_AS_ERRORS=0 COMPILER_INDEX_STORE_ENABLE=NO'
2026-08-13 13:28:47:159 i [Xcode] CODE_SIGN_IDENTITY = iPhone Developer
2026-08-13 13:28:47:159 i [Xcode] DEVELOPMENT_TEAM = <TEAM_ID>
2026-08-13 13:28:47:160 i [Xcode] IPHONEOS_DEPLOYMENT_TARGET = 17.5
2026-08-13 13:28:47:598 i [Xcode] note: Using codesigning identity override: iPhone Developer

[Signing identity resolution]
2026-08-13 13:28:51:483 i [Xcode] Matching identities included multiple entries with the display name "Apple Development: <SIGNER>".

[Failure]
2026-08-13 13:28:51:488 i [Xcode] ** TEST BUILD FAILED **
2026-08-13 13:28:51:489 i [Xcode] Run custom shell script 'Embed app icon into Runner.app'
2026-08-13 13:28:51:496 i [Xcode] (3 failures)
2026-08-13 13:28:56:871 i [Xcode] Recovery Suggestion: Failed to verify code signature of <DEVICE_PATH> : 0xe8008001 (An unknown error has occurred.)
2026-08-13 13:28:56:876 i [Xcode] FunctionName = "+[MICodeSigningVerifier _validateSignatureAndCopyInfoForURL:withOptions:error:]";
2026-08-13 13:28:56:876 i [Xcode] LegacyErrorString = ApplicationVerificationFailed;
2026-08-13 13:28:56:879 i [Xcode] Failed to install the app on the device.
2026-08-13 13:29:10:051 e [Xcode] xcodebuild exited with code '65' and signal 'null'
2026-08-13 13:29:10:052 d [XCUITestDriver] Error: xcodebuild failed with code 65.
```

--- a/Scripts/embed-runner-icon.sh
+++ b/Scripts/embed-runner-icon.sh
@@ -36,12 +36,32 @@ if [ ${#ICONS[@]} -eq 0 ]; then
     exit 0
 fi
 
-cp -f "${ICONS[@]}" "$RUNNER_APP/"
+PLIST="$RUNNER_APP/Info.plist"
+
+# Everything below mutates a bundle Xcode already signed, so keep enough state
+# to undo it: a failed re-sign must not leave an uninstallable runner behind.
+PLIST_BACKUP=$(mktemp)
+CERT_DIR=$(mktemp -d)
+trap 'rm -rf "$PLIST_BACKUP" "$CERT_DIR"' EXIT
+cp "$PLIST" "$PLIST_BACKUP"
+ADDED_FILES=()
+
+restore_bundle() {
+    cp -f "$PLIST_BACKUP" "$PLIST"
+    if [ ${#ADDED_FILES[@]} -gt 0 ]; then
+        rm -f "${ADDED_FILES[@]}"
+    fi
+}
+
+for icon in "${ICONS[@]}"; do
+    dest="$RUNNER_APP/$(basename "$icon")"
+    [ -e "$dest" ] || ADDED_FILES+=("$dest")
+    cp -f "$icon" "$dest"
+done
 if [ -f "$XCTEST/Assets.car" ]; then
+    [ -e "$RUNNER_APP/Assets.car" ] || ADDED_FILES+=("$RUNNER_APP/Assets.car")
     cp -f "$XCTEST/Assets.car" "$RUNNER_APP/"
 fi
-
-PLIST="$RUNNER_APP/Info.plist"
 /usr/libexec/PlistBuddy -c "Delete :CFBundleIcons" "$PLIST" 2>/dev/null || true
 /usr/libexec/PlistBuddy -c "Delete :CFBundleIcons~ipad" "$PLIST" 2>/dev/null || true
 
@@ -58,29 +78,40 @@ PLIST="$RUNNER_APP/Info.plist"
 /usr/libexec/PlistBuddy -c "Add :CFBundleIcons~ipad:CFBundlePrimaryIcon:CFBundleIconFiles:0 string AppIcon60x60" "$PLIST"
 /usr/libexec/PlistBuddy -c "Add :CFBundleIcons~ipad:CFBundlePrimaryIcon:CFBundleIconFiles:1 string AppIcon76x76" "$PLIST"
 
-# Re-codesign since we modified the bundle after Xcode signed it.
-# In a scheme post-action context Xcode's CODE_SIGN_* env vars are not exposed,
-# so discover the existing signing identity from the already-signed bundle.
+# Re-codesign since we modified the bundle after Xcode signed it. Scheme
+# post-actions never receive EXPANDED_CODE_SIGN_IDENTITY, so the signer has to
+# be recovered from the already-signed bundle.
 if [ -d "$RUNNER_APP/_CodeSignature" ]; then
     # Capture the signature info once. Piping codesign straight into
     # `awk ... exit` makes awk close the pipe early, killing codesign with
     # SIGPIPE -- which `set -o pipefail` turns into a fatal error. That trips
     # only when an Authority line exists, i.e. on every real-device build.
     SIGN_INFO=$(codesign -dvv "$RUNNER_APP" 2>&1 || true)
-    EXISTING_IDENT="${EXPANDED_CODE_SIGN_IDENTITY:-}"
-    if [ -z "$EXISTING_IDENT" ]; then
-        EXISTING_IDENT=$(awk -F'=' '/^Authority/ {print $2; exit}' <<< "$SIGN_INFO")
-    fi
-    # Simulator builds are ad-hoc signed: there is no Authority line, but the
-    # bundle can still be re-signed ad-hoc with an identity of "-".
-    if [ -z "$EXISTING_IDENT" ] && grep -q '^Signature=adhoc' <<< "$SIGN_INFO"; then
+
+    if grep -q '^Signature=adhoc' <<< "$SIGN_INFO"; then
+        # Simulator builds are ad-hoc signed; re-sign them ad-hoc.
         EXISTING_IDENT="-"
-    fi
-    if [ -n "$EXISTING_IDENT" ]; then
-        codesign --force --sign "$EXISTING_IDENT" \
-                 --preserve-metadata=identifier,entitlements "$RUNNER_APP"
     else
-        echo "warning: bundle is signed but no identity discovered; signature will be invalid"
+        # Identify the signer by leaf-certificate SHA-1, never by Authority
+        # display name: duplicate identities can share a name, and codesign
+        # then rejects the name as ambiguous (Apple TN3161).
+        EXISTING_IDENT=""
+        if codesign -d --extract-certificates="$CERT_DIR/leaf" "$RUNNER_APP" 2>/dev/null \
+                && [ -f "$CERT_DIR/leaf0" ]; then
+            EXISTING_IDENT=$(shasum -a 1 "$CERT_DIR/leaf0" | awk '{print toupper($1)}')
+        fi
+        if [ -z "$EXISTING_IDENT" ]; then
+            EXISTING_IDENT=$(awk -F'=' '/^Authority/ {print $2; exit}' <<< "$SIGN_INFO")
+        fi
+    fi
+
+    if [ -z "$EXISTING_IDENT" ] || ! codesign --force --sign "$EXISTING_IDENT" \
+            --preserve-metadata=identifier,entitlements "$RUNNER_APP"; then
+        # A modified-but-stale-signed bundle fails to install, which would break
+        # the session over a cosmetic icon. Undo instead.
+        restore_bundle
+        echo "warning: could not re-sign $RUNNER_APP; skipped icon embed"
+        exit 0
     fi
 fi
 

